src/osp/Active/SysNewton.cpp

```diff
@@ -352,32 +362,84 @@ ACompNwtWorld* SysNewton::try_get_physics_world(ActiveScene &rScene)
 }
 
 std::pair<ActiveEnt, ACompNwtBody*> SysNewton::find_rigidbody_ancestor(
-        ActiveScene& rScene, ActiveEnt ent)
+        ActiveScene& rScene, ActiveEnt ent, Matrix4* pTransform)
 {
     ActiveEnt prevEnt, currEnt = ent;
-    ACompHierarchy *currHeir = nullptr;
+    ACompHierarchy *currHier = nullptr;
 
     do
     {
+       currHier = rScene.get_registry().try_get<ACompHierarchy>(currEnt);
 
-       currHeir = rScene.get_registry().try_get<ACompHierarchy>(currEnt);
-
-        if (!currHeir)
+        if (!currHier)
         {
             return {entt::null, nullptr};
         }
+        
+        // If transform argument was passed, record the local transformation of
+        // the current node relative to its parent
+        if (pTransform && (currHier->m_level > gc_heir_physics_level))
+        {
+            Matrix4 localTransform = rScene.reg_get<ACompTransform>(currEnt).m_transform;
+            *pTransform = localTransform * (*pTransform);
+        }
 
         prevEnt = currEnt;
-        currEnt = currHeir->m_parent;
+        currEnt = currHier->m_parent;
     }
-    while (currHeir->m_level != gc_heir_physics_level);
+    while (currHier->m_level != gc_heir_physics_level);
 
-    auto *body = rScene.get_registry().try_get<ACompNwtBody>(prevEnt);
+    auto *body = rScene.get_registry().try_get<ACompRigidBody_t>(prevEnt);
 
     return {prevEnt, body};
 
 }
 
+std::pair<ACompNwtBody*, ACompTransform*> SysNewton::try_get_or_find_rigidbody_parent(
+    ActiveScene& rScene, ActiveEnt childEntity, ActiveEnt& rParentRigidBody)
+{
+    ACompNwtBody *compRb;
+    ACompTransform *compTf;
```

Comment:
No reason to declare these variables here.

Declare them where they are defined, in both places.

Reply:
Then return them inside the scope they're declared, instead of a single return at the end.